### PR TITLE
Sort the best traces as we go using insertion sort instead of sorting the whole array each time

### PR DESCRIPTION
**Pull request by romac**
See this thread for background information: https://github.com/informalsystems/quint/pull/1465#discussion_r1693936729

This might be completely overkill if `n-traces` is in practice less than a few thousands, but it was fun to write :) No worries if you'd rather not merge it because of that.

---

- [x] Tests added for any new code
- [ ] Documentation added for any new functionality
- [ ] Entries added to the respective `CHANGELOG.md` for any new functionality
- [ ] Feature table on [`README.md`](../README.md#roadmap) updated for any listed functionality

### CODE REVIEW

**Comment by romac** [comment]
Here is a [benchmark](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJWRFFhbWg3TWg2ZWpvV3hIN3BRWk8iLCJjb2RlIjoidmFyIHNvcnRlZCA9IFtdXG5cbmZvciAodmFyIGkgPSAwOyBpIDwgREFUQS5sZW5ndGg7IGkgKz0gMSkge1xuICBzb3J0ZWQucHVzaChEQVRBW2ldKTtcbiAgc29ydGVkLnNvcnQoKGEsIGIpID0-IGEgLSBiKTtcbn0iLCJuYW1lIjoiTmF0aXZlIHNvcnQiLCJkZXBlbmRlbmNpZXMiOltdfSx7ImlkIjoiS1ZKLUg3aWp1UG04SkVsRUM5QW1GIiwiY29kZSI6IiAgZnVuY3Rpb24gaW5zZXJ0aW9uU29ydChiZXN0VHJhY2VzLCB0cmFjZSkge1xuICAgIGlmIChiZXN0VHJhY2VzLmxlbmd0aCA9PT0gMCkge1xuICAgICAgYmVzdFRyYWNlcy5wdXNoKHRyYWNlKTtcbiAgICAgIHJldHVybjtcbiAgICB9XG5cbiAgICBjb25zdCBpbnNlcnRpb25JbmRleCA9ICgoKSA9PiB7XG4gICAgICB2YXIgbG93ID0gMDtcbiAgICAgIHZhciBoaWdoID0gYmVzdFRyYWNlcy5sZW5ndGg7XG5cbiAgICAgIHdoaWxlIChsb3cgPCBoaWdoKSB7XG4gICAgICAgIGNvbnN0IG1pZCA9IChsb3cgKyBoaWdoKSA-Pj4gMTtcbiAgICAgICAgaWYgKHRyYWNlIC0gYmVzdFRyYWNlc1ttaWRdIDwgMCkge1xuICAgICAgICAgIGhpZ2ggPSBtaWQ7XG4gICAgICAgIH0gZWxzZSB7XG4gICAgICAgICAgbG93ID0gbWlkICsgMTtcbiAgICAgICAgfVxuICAgICAgfVxuXG4gICAgICByZXR1cm4gbG93O1xuICAgIH0pKCk7XG5cbiAgICBiZXN0VHJhY2VzLnNwbGljZShpbnNlcnRpb25JbmRleCwgMCwgdHJhY2UpO1xufVxuXG52YXIgc29ydGVkID0gW11cblxuZm9yICh2YXIgaSA9IDA7IGkgPCBEQVRBLmxlbmd0aDsgaSArPSAxKSB7XG4gIGluc2VydGlvblNvcnQoc29ydGVkLCBpKTtcbn0iLCJuYW1lIjoiSW5zZXJ0aW9uIHNvcnQiLCJkZXBlbmRlbmNpZXMiOltdfV0sImNvbmZpZyI6eyJuYW1lIjoiQmFzaWMgZXhhbXBsZSIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6InJldHVybiBbLi4uQXJyYXkoMTAwMCkua2V5cygpXS5tYXAoeCA9PiBNYXRoLnJvdW5kKE1hdGgucmFuZG9tKCkgKiAxMDAwKSkifX0) to compare the two approaches.

**Comment by romac** [comment]
@bugarela Let me know if you want to move forward with this, and I will add a changelog entry. Otherwise, no problem :)

**Comment by bugarela** [comment]
> @bugarela Let me know if you want to move forward with this, and I will add a changelog entry. Otherwise, no problem :)

I absolutely do! I haven't reviewed it since it is marked as a draft, but I can do it. Thank you for doing this!

**Comment by romac** [comment]
Ok great, I'll let you take care of the changelog then :)

File: quint/src/util.ts [hunk 1]
```diff
@@ -68,3 +68,43 @@ export function findMap<X, Y>(xs: Iterable<X>, f: (x: X) => Maybe<Y>): Maybe<Y>
...
+    array.push(item)
+    return
+  }
+
+  const index = findInsertionIndex(array, x => cmp(x, item) > 0)
+  array.splice(index, 0, item)
+}
+
+/** Find the `index` in `array` where the predicate `pred(array[index])` is true,
+ * but `pred(array[index+1])` is false.
```
Comment (romac) [inline quint/src/util.ts:84]: This comment comes from a former version of this function and should be updated:

```suggestion
/** Find the `index` in `array` where `pred(array[index])` is true.
```

File: quint/src/util.ts [hunk 2]
```diff
@@ -68,3 +68,47 @@ export function findMap<X, Y>(xs: Iterable<X>, f: (x: X) => Maybe<Y>): Maybe<Y>
...
+  if (array.length === 0) {
+    array.push(item)
+    return
+  }
+
+  const index = findInsertionIndex(array, x => cmp(x, item) > 0)
+  array.splice(index, 0, item)
+}
+
+/** Find the `index` in `array` where the predicate `pred(array[index])` is true.
```
Comment (bugarela) [inline quint/src/util.ts:88]:
```suggestion
/** Find the first `index` in `array` where the predicate `pred(array[index])` is true.
```

**Pull request merged**